spectre_analyses/tasks.py

```diff
@@ -0,0 +1,159 @@
+limitations under the License.
+"""
+from contextlib import contextmanager
+from functools import partial
+import json
+import os
+import pickle
+import re
+import shutil
+import sys
+
+import celery
+from celery.utils.log import get_task_logger
+
+import spdata.reader
+import spdata.types as ty
+
+from spectre_analyses.celery import app
+import matlab_hooks as mh
+
+
+FILESYSTEM_ROOT = os.path.abspath(os.sep)
+DATA_ROOT = os.path.join(FILESYSTEM_ROOT, 'data')
+STATUS_PATHS = {
+    'all': FILESYSTEM_ROOT,
+    'done': DATA_ROOT,
+    'processing': os.path.join(FILESYSTEM_ROOT, 'temp'),
+    'failed': os.path.join(FILESYSTEM_ROOT, 'failed')
+}
+Name = str
+Path = str
+DISALLOWED_CHARACTERS = "[^a-zA-Z0-9_-]"
+
+
+def _data_path(dataset_name: Name) -> Path:
+    dataset_name = re.sub(DISALLOWED_CHARACTERS, '_', dataset_name)
+    return os.path.join(DATA_ROOT, dataset_name, 'text_data', 'data.txt')
+
+
+def _load_data(dataset_name) -> ty.Dataset:
+    path = _data_path(dataset_name)
+    with open(path) as infile:
+        return spdata.reader.load_txt(infile)
+
+
+@contextmanager
+def _open_analysis(dataset_name: str, algorithm_name: str, analysis_name: str):
+    analysis_root = os.path.join(
+        STATUS_PATHS['processing'],
+        dataset_name,
+        algorithm_name,
+        analysis_name,
+    )
+    os.makedirs(analysis_root)
+    try:
+        yield analysis_root
+        dest_root = os.path.join(
+            STATUS_PATHS['done'],
+            dataset_name,
+            algorithm_name,
+            analysis_name
+        )
+    except Exception as ex:
+        dest_root = os.path.join(
+            STATUS_PATHS['failed'],
+            dataset_name,
+            algorithm_name,
+            analysis_name
+        )
+        raise RuntimeError() from ex
+    finally:
+        shutil.move(analysis_root, dest_root)
+
+
+def _simply_typed(result: mh.DivikResult):
+    result = result._asdict()
+    result['centroids'] = result['centroids'].tolist()
+    result['partition'] = result['partition'].tolist()
+    result['quality'] = float(result['quality'])
+    result['filters'] = {
+        key: result['filters'][key].tolist() for key in result['filters']
+    }
+    result['thresholds'] = {
+        key: float(result['thresholds'][key]) for key in result['thresholds']
+    }
+    result['merged'] = result['merged'].tolist()
+    result['subregions'] = [
+        _simply_typed(subregion) if subregion is not None else None
+        for subregion in result['subregions']
+    ]
+    return result
+
+
+def _notify(task, status):
+    task.update_state(state=status)
+    # Line below updates the status in Celery Flower.
+    # It is disabled since Flower disables TERMINATE button for custom state.
+    #task.send_event('task-' + status.lower().replace(' ', '_'))
+
+
+@contextmanager
+def _status_notifier(task: celery.Task):
+    old_outs = sys.stdout, sys.stderr
+    rlevel = task.app.conf.worker_redirect_stdouts_level
```

Comment:
rename to `redirect_level`?

Reply:
Pinned to next task: https://trello.com/c/lV7a9dVe/415-re-style-divik-inputs-form